README.md
# brandaniethical

A package for green consumers and investors to quickly get a view of fashion brands’ animal policy, rate, and animal material usage. and check trendy tickers esg from yahoo finance and esg api

### Function1 - brandaniethical(brandlist)
The function provides a quick look at brandlists animal usage information, standards, and scores.
animal material ethical information include usage of fur, angora, down feather, shearling, karakul, exotic animal skin and hair, wool use including ‘mulesing’, and leather. Also check whether brands use these material align with standards.
In vignette file, a presentation demo with class of popular fashion brand is provided by retrieving 6 popoluar footwear and apparel brands' tickers yahoo finance.  
Skills:
Data Acquisition:
Import data from json.
Web scrape data from public websites: https://goodonyou.eco/
Write an API client for an API and/or functions associated with API interaction.
Handle, parse, and transform JSON and XML.
Data Cleaning, Transformation, and Organization:
Use data wrangling to transform data into a dataframe ready for green consumers reference and analysis.
Use loops processes.
Use functions and functional programming to export repetitive or difficult tasks.
Handle and process strings and use regular expressions.
Documentation and Presentation:
Provide and document useful functions as part of a Python package.
Capture error and exceptions
Use class
Use Jupyter Notebook to generate a vignette on how to use your package functions.

    """
    Get fashion brands' animal ethical usage information. provide information of whether the brand use fur, angora, down feather, shearling, karakul, exotic animal skin and hair, wool use including ‘mulesing’, and leather 
    Also check whether brands use these material align with standards.
    
    Parameters
    ----------
    brandlist : list
      A list of brands name that consumers want to know their condition of animal usage
    
    Returns
    -------
    pandas.dataframe
      The dataframe of brands ethical consumption of animal material which include usage of fur, angora, down feather, shearling, karakul, exotic animal skin and hair, wool use including ‘mulesing’, and leather. Also check whether brands use these material align with standards.
      
    Examples
    --------
    >>> from brandaniethical import brandaniethical 
    >>> brandaniethical.brandaniethical(['cos','theory','lululemon','nike', 'skechers'])
	company	animal_score	update	fur	angora	leather	wool	shearling	karakul	down	feather	exotic animal hair	exotic animal skin	down_RDS	wool_mulesing
	0	cos	3/5	July 2020	0	0	1	1	0	0	1	0	1	0	1	1
	1	theory	2/5	December 2019	0	0	1	1	0	0	1	0	1	0	0	0
	2	lululemon	2/5	July 2020	0	0	1	1	0	0	1	0	1	0	1	0
	3	nike	2/5	July 2020	0	0	1	1	0	0	1	0	1	0	0	0
	4	skechers	2/5	August 2020	0	0	1	1	0	0	0	0	0	0	0	0
    
    """
    
### Function2: trendyesg(region, start=5)
Get trendy tickers' company information from Yahoo Finance api and return their ESG score, environment score, social score, and governance score from ESG api. Users can check trending tickets across market regions, sort by top mktprice, start from specified rank of top marketing price tickers. uplimit = 20
region = ['US', 'BR', 'AU', 'CA', 'FR', 'DE', 'HK', 'IN', 'IT', 'ES', 'GB', 'SG']
Users can make fuzzy search of those trendy tickers by discarding dollar symbol such as usd, cad from trendy tickers search results. For example, shib-usd and shib-cad are both searched with 'shib' keyword, and the esg api only returns the stem ticker symbol.
Also, due to the rate limit of esg api, I set trendy tickers search results limited in 5 tickers. but users can start at any point of trendy tickers list sorted by market prices, the default is start from the 5th higest market price trendy tickers and return 5-10 trendy tickers esg scores.
Skills:
Data Acquisition:
Import data from json.
Use APIs to obtain data. - esg api + yahoo finance api
Yahoo finance API: https://rapidapi.com/apidojo/api/yh-finance/
ESG API document: https://www.esgenterprise.com/esg-enterprise-data-api-services/
Write an API client for an API and/or functions associated with API interaction.
Handle, parse, and transform JSON and XML.
Data Cleaning, Transformation, and Organization:
Use data wrangling to transform data into a dataframe ready for analysis.
Use loops processes.
Use functions and functional programming to export repetitive or difficult tasks.
Handle and process strings and use regular expressions.
Documentation and Presentation:
Provide and document useful functions as part of a Python package.
capture error
Use Jupyter Notebook to generate a vignette on how to use your package functions.

    
    """
    Get region's trendy tickers' esg score based on given region and regular rank of market price.

    Parameters
    ----------
    region : string
      A string that limit the region of market consumers are interested in, which include US, BR, AU, CA, FR, DE, HK, IN, IT, ES, GB, and SG.
    start : int
      A int that indicate the ranking .

    Returns
    -------
    pandas.dataframe
      The dataframe of trendy tickers' esg, environmental, social,governance score, and its ticker based on consumers interested region and the lowest ranking consumers want to start to look at

    Examples
    --------
    >>> from brandaniethical import trendyesg 
    >>> trendyesg(region='US',start =8)
	company_name	env_score	soc_score	gov_score	total	ticker
	0	Medtronic plc	500	304	310	1114	MDT
	1	Pfizer Inc.	510	300	305	1115	PFE
	2	Roblox Corporation	245	304	220	769	RBLX
	3	Lemonade, Inc.	520	303	300	1123	LMND
	4	Nucor Corporation	251	326	210	787	NUE
    
    """
    

## Installation

```bash
$ pip install brandaniethical
```

## Usage

- TODO

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`brandaniethical` was created by Yang Hu. It is licensed under the terms of the MIT license.

## Credits

`brandaniethical` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (15/12/2021)

- First release of `brandaniethical`!

tests/test_brandaniethical.py
import re
import requests
from bs4 import BeautifulSoup
import pandas as pd
from requests.exceptions import HTTPError
import requests.auth
import os
from brandaniethical.brandaniethical import brandaniethical
from brandaniethical.brandaniethical import trendyesg
from pandas._testing import assert_frame_equal

#function1 test
df1 = pd.DataFrame(
{'company':'adidas',
 'animal_score':'3/5',
 'update':'July 2020',
 'fur':0,
 'angora':0,
 'leather':1,
 'wool':1,
 'shearling':0,
 'karakul':0,
 'down':0,
 'feather':0,
 'exotic animal hair':1,
 'exotic animal skin':0,
 'down_RDS':0,
 'wool_mulesing':1
},index=[0])

df2 = pd.DataFrame(
{'company':'nike',
 'animal_score':'2/5',
 'update':'July 2020',
 'fur':0,
 'angora':0,
 'leather':1,
 'wool':1,
 'shearling':0,
 'karakul':0,
 'down':1,
 'feather':0,
 'exotic animal hair':1,
 'exotic animal skin':0,
 'down_RDS':0,
 'wool_mulesing':0
},index=[0])

def test_fashion_animaterial():
    example = ['adidas']
    expected = df1
    result = brandaniethical(example)
    assert_frame_equal(df1, result)
    

def test_uppercase_brand():
    example2 = ['NIKE']
    expected = df2
    result = brandaniethical(example2)
    assert_frame_equal(df2, result)

def test_invalid_brand():
    example3 = ['a']
    result = brandaniethical(example3)
    assert isinstance(result, type(None))
    
    
    
#function2 test
import pytest
def test_invalid_region():
    region = 'ak'
    with pytest.raises(AssertionError):
        trendyesg(region, start=5)


def test_invalid_start():
    start = 'ak'
    with pytest.raises(AssertionError):
        trendyesg(region='US', start=start)


src/brandaniethical/brandaniethical.py
import re
import requests
from bs4 import BeautifulSoup
import pandas as pd
from requests.exceptions import HTTPError
import requests.auth
import os
# function 1: get fashion brands' animal ethical info 
def brandaniethical(brandlist):
    
    """
    Get fashion brands' animal ethical usage information. provide information of whether the brand use fur, angora, down feather, shearling, karakul, exotic animal skin and hair, wool use including ‘mulesing’, and leather 
    Also check whether brands use these material align with standards.
    
    Parameters
    ----------
    brandlist : list
      A list of brands name that consumers want to know their condition of animal usage
    
    Returns
    -------
    pandas.dataframe
      The dataframe of brands ethical consumption of animal material which include usage of fur, angora, down feather, shearling, karakul, exotic animal skin and hair, wool use including ‘mulesing’, and leather. Also check whether brands use these material align with standards.
      
    Examples
    --------
    >>> from brandaniethical import brandaniethical 
    >>> brandaniethical.brandaniethical(['cos','theory','lululemon','nike', 'skechers'])
	company	animal_score	update	fur	angora	leather	wool	shearling	karakul	down	feather	exotic animal hair	exotic animal skin	down_RDS	wool_mulesing
0	cos	3/5	July 2020	0	0	1	1	0	0	1	0	1	0	1	1
1	theory	2/5	December 2019	0	0	1	1	0	0	1	0	1	0	0	0
2	lululemon	2/5	July 2020	0	0	1	1	0	0	1	0	1	0	1	0
3	nike	2/5	July 2020	0	0	1	1	0	0	1	0	1	0	0	0
4	skechers	2/5	August 2020	0	0	1	1	0	0	0	0	0	0	0	0
    """
    assert isinstance(brandlist, list), f"This function only works on list but brandlist={brandlist}."
    for b in brandlist:
        assert isinstance(b, str), f"This function only works on string but brand={b}."
    brandlist = [b.lower() for b in brandlist]
    ani_reg = r'[^.]* animal [^.]*\.'
    use_reg = r'[^.]* (uses) [^.]*\.'
    date_reg = r'(Last Updated: )(\w*\s\d...)'
    web = 'https://directory.goodonyou.eco/'
    animal_score = []
    ani_rate= []
    company = []
    update = []
    def brandani(brandlist):
        df_gfy=pd.DataFrame()
        for brand in brandlist:
            goypage = requests.get(web +'brand/'+ brand)
            soup = BeautifulSoup(goypage.content, 'html.parser')
            try:
                animal_score.append(str(soup.find_all("span", {"class": "StyledText-sc-1sadyjn-0 ccIhDL"})[2].string))
                ani_rate.append([string for string in soup.stripped_strings if re.search(ani_reg, string)][0])
                update.append([string for string in soup.stripped_strings if re.search(date_reg, string)][0]) 
            except (IndexError, TypeError, NameError, KeyError):
                print(f'please check company name of {brand}')
                pass
            else:
                df_gfy['company'] = pd.Series(brandlist)
                df_gfy['animal_score'] = pd.Series(animal_score)
                df_gfy['ani_rate'] = pd.Series(ani_rate)
                df_gfy['update'] = pd.Series(update)
                pd.set_option('display.max_colwidth', None)
        return df_gfy
    global dfydf
    dfydf = brandani(brandlist)
    #check animal material usage
    #data wrangling
    try:
        
        dfydf['ani_material'] = [re.search(use_reg, i).group(0) for i in dfydf['ani_rate']]

    except (IndexError, TypeError, NameError, KeyError):
        print(f'no animal material information')
        

    else:
        dfydf['update'] = [re.search(r'\w*\s\d...', i).group(0) for i in dfydf['update']]
        dfydf['animal_score'] = [re.sub(r' out of ', '/', i) for i in dfydf['animal_score']]
        fabric = ['fur','angora', 'leather', 'wool', 'shearling', 'karakul', 'down', 'feather', 'exotic animal hair', 'exotic animal skin']
        for f in fabric:
            dfydf[f]=dfydf['ani_material'].str.contains(f)
        #check if use down whether down accredited by the Responsible Down Standar
        RDS_reg = 'It uses down accredited by the Responsible Down Standar'
        dfydf['down_RDS'] = dfydf['ani_rate'].str.contains(RDS_reg)
        dfydf.loc[dfydf['down_RDS']==True,'down'] = True
        #check if use wool whether wool includes mulesing
        dfydf['wool_mulesing'] = dfydf['ani_rate'].str.contains('mules')
        dfydf.loc[dfydf['wool_mulesing']==True,'wool'] = True
        df_aniusage = dfydf.loc[:,dfydf.columns!='ani_material']
        df_material = df_aniusage.replace(True,1)
        df_material = df_material.replace(False,0)
        
        return df_material.loc[:,df_material.columns!='ani_rate']
       
       
    
# function 2: get region trendytickers esg score       
import requests
from requests.exceptions import HTTPError
import requests.auth
import re
import pandas as pd
import os
regionlist = ['US', 'BR', 'AU', 'CA', 'FR', 'DE', 'HK', 'IN', 'IT', 'ES', 'GB', 'SG']
YAHOOF_KEY = os.getenv('YAHOOF_KEY')
app_key = os.getenv('ESG_KEY')

def trendyesg(region, start=5):
    """
    Get region's trendy tickers' esg score based on given region and regular rank of market price.

    Parameters
    ----------
    region : string
      A string that limit the region of market consumers are interested in, which include US, BR, AU, CA, FR, DE, HK, IN, IT, ES, GB, and SG.
    start : int
      A int that indicate the ranking .

    Returns
    -------
    pandas.dataframe
      The dataframe of trendy tickers' esg, environmental, social,governance score, and its ticker based on consumers interested region and the lowest ranking consumers want to start to look at

    Examples
    --------
    >>> from brandaniethical import trendyesg 
    >>> trendyesg(region='US',start =8)
	company_name	env_score	soc_score	gov_score	total	ticker
0	Medtronic plc	500	304	310	1114	MDT
1	Pfizer Inc.	510	300	305	1115	PFE
2	Roblox Corporation	245	304	220	769	RBLX
3	Lemonade, Inc.	520	303	300	1123	LMND
4	Nucor Corporation	251	326	210	787	NUE
    """
    assert region in regionlist, f"This function only works on {regionlist} but region={region}."
    assert isinstance(start, int), f"This function only works on int but start={start}."
    def get_trendycom(region, start=5):
        url = "https://yh-finance.p.rapidapi.com/market/get-trending-tickers"
        def trendytick(region):
            querystring = {"region":region}
            try:
                headers = {
                    'x-rapidapi-host': "yh-finance.p.rapidapi.com",
                    'x-rapidapi-key': YAHOOF_KEY
                            }
                response = requests.request("GET", url, headers=headers, params=querystring)
                response.raise_for_status()

            except HTTPError as http_err:
                print(f'HTTP error occurred: {http_err}')
            except Exception as err: 
                print(f'Other error occurred: {err}')
            return response.json()

        cominfolist = trendytick(region)

        region = []
        shortName = []
        regularMarketPrice = []
        symbol=[]
        trendtick_df = pd.DataFrame()
    
        for i in cominfolist['finance']['result'][0]['quotes']:   
            shortName.append(i['shortName'])
            regularMarketPrice.append(i['regularMarketPrice'])
            symbol.append(i['symbol'])

        trendtick_df['shortName'] = pd.Series(shortName)
        trendtick_df['regularMarketPrice'] = pd.Series(regularMarketPrice)
        trendtick_df['symbol'] = pd.Series(symbol)
        trendtick_df['maxprice_rank'] = trendtick_df['regularMarketPrice'].rank(method='max')

        return trendtick_df.sort_values(by=['maxprice_rank']).iloc[start:start+5]

    #get trendy tickers outcome
    tdf = get_trendycom(region, start=5)
    #make general search of ticker e.g.
    ttlist = [i for i in tdf['symbol']]
    reg_tick = r'(^.+?)-|(^.*)'
    for tick in ttlist:
        ticklist=[]
        try:
            ticklist = [re.match(reg_tick,tick).group(1) for tick in ttlist]
            ticklist.extend([re.match(reg_tick,tick).group(2) for tick in ttlist])
            ticklist = set(ticklist)
        except Exception as e:
            print(f'{e}: the ticker is not supported')
    ticklist = list(filter(None, ticklist))

    base_url = 'https://tf689y3hbj.execute-api.us-east-1.amazonaws.com/prod/authorization/'
    df = pd.DataFrame()
    company_name=[]
    env_score=[]
    soc_score=[]
    gov_score=[]
    total=[]
    region=[]
    ticker=[]
    for brand_tick in ticklist:
        try:
            response_esg = requests.get(base_url+'search?',params={'q':brand_tick,'token':app_key})
            response_esg.raise_for_status()
            company_name.append(response_esg.json()[0]['company_name'])
            env_score.append(response_esg.json()[0]['environment_score'])
            soc_score.append(response_esg.json()[0]['social_score'])
            gov_score.append(response_esg.json()[0]['governance_score'])
            total.append(response_esg.json()[0]['total'])
            ticker.append(brand_tick)
        except IndexError as ind_err:
            company_name.append(brand_tick)
            env_score.append('NA')
            soc_score.append('NA')
            gov_score.append('NA')
            total.append('NA')
            ticker.append(brand_tick)
            print(f"the company's esg not in database")
        except HTTPError as http_err:
            print(f'HTTP error occurred: {http_err}')
        except Exception as err: 
            print(f'Other error occurred: {err}')
    df['company_name'] = pd.Series(company_name)
    df['env_score'] = pd.Series(env_score)
    df['soc_score'] = pd.Series(soc_score)
    df['gov_score'] = pd.Series(gov_score)
    df['total'] = pd.Series(total)
    df['ticker'] = pd.Series(ticker)
    return df